packages/electron/src/electron-main/electron-application.ts

```diff
@@ -0,0 +1,429 @@
+        electronWindow.loadURL(uri.toString(true));
+        return electronWindow;
+    }
+
+    async openWindowWithWorkspace(workspace: string): Promise<BrowserWindow> {
+        const uri = (await this.createWindowUri()).withFragment(workspace);
+        const electronWindow = await this.createWindow(this.getBrowserWindowOptions());
+        electronWindow.loadURL(uri.toString(true));
+        return electronWindow;
+    }
+
+    /**
+     * "Gently" close all windows, application will not stop if a `beforeunload` handler returns `false`.
+     */
+    requestStop(): void {
+        app.quit();
+    }
+
+    isStopping(): boolean {
+        return typeof this.stoppingTask !== 'undefined';
+    }
+
+    protected async handleMainCommand(params: ExecutionParams, options: MainCommandOptions): Promise<void> {
+        if (typeof options.file === 'undefined') {
+            await this.openDefaultWindow();
+        } else {
+            await this.openWindowWithWorkspace(realpathSync(path.resolve(params.cwd, options.file)));
+        }
+    }
+
+    protected async createWindowUri(): Promise<URI> {
+        return FileUri.create(this.globals.THEIA_FRONTEND_HTML_PATH)
+            .withQuery(`port=${this.backendPort}`);
+    }
+
+    protected getBrowserWindowOptions(): BrowserWindowConstructorOptions {
+        let windowState: BrowserWindowConstructorOptions | undefined = this.electronStore.get('windowstate', undefined);
+        if (typeof windowState === 'undefined') {
+            windowState = this.getDefaultWindowState();
+        }
+        return {
+            ...windowState,
+            show: false,
+            title: this.globals.THEIA_APPLICATION_NAME,
+            minWidth: 200,
+            minHeight: 120,
+        };
+    }
+
+    protected getDefaultWindowState(): BrowserWindowConstructorOptions {
+        // The `screen` API must be required when the application is ready.
+        // See: https://electronjs.org/docs/api/screen#screen
+        const { screen } = require('electron');
+        // We must center by hand because \`browserWindow.center()\` fails on multi-screen setups
+        // See: https://github.com/electron/electron/issues/3490
+        const { bounds } = screen.getDisplayNearestPoint(screen.getCursorScreenPoint());
+        const height = Math.floor(bounds.height * (2 / 3));
+        const width = Math.floor(bounds.width * (2 / 3));
+        const y = Math.floor(bounds.y + (bounds.height - height) / 2);
+        const x = Math.floor(bounds.x + (bounds.width - width) / 2);
+        return { width, height, x, y };
+    }
+
+    /**
+     * Prevent opening links into new electron browser windows by default.
+     */
+    protected attachWebContentsNewWindow(electronWindow: BrowserWindow): void {
+        electronWindow.webContents.on('new-window', (event, url) => {
+            event.preventDefault();
+            shell.openExternal(url);
+        });
+    }
```

Comment:
@marechal-p, do we still need this code or is it a leftover?

Originally, it was added [here](https://github.com/eclipse-theia/theia/commit/737fb8f86d39368d591ce3f790c22cb867f378c3#diff-e6697ec7c0f7ddbbfd9e8365baacf132R142-R147), but if I remove it, I can still open URLs from a MD file in my OS browser. For externals, we use `shell.openExternal(url)` anyway. What do you think?

Reply:
It is a "safe" mesure, I really think no new Electron Browser Window should be opened when a link is clicked upon.

Reply:
OK, thanks for confirming. I am going to remove it in #8076. We should make sure that 
 - all of the Theia extensions correctly use the `WindowService` and not `window.open`,
 - but we should not customize the default electron behavior and add custom, unused event handlers just because we want to be "safe". Downstream projects/extensions can decide what they want to do or how they want to mess with the security.

Thoughts?
// CC: @marcdumais-work, @akosyakov

Reply:
This behavior is triggered when you click an HTML link.

Reply:
How can I reproduce it? I tried from a markdown editor, with the `Open URL` command for a preview, and from the _Mini Browser_. All opened up in my OS browser as an external after removing the corresponding code. Thanks!

Reply:
Tried what you mentioned, it seems like something changed and without this code no new electron window is opened indeed. It wasn't the case before, I'm certain. But using DOM's API such as `window.open(...)` it opens a new Electron window. I still don't see why someone would want to open an Electron window like this, it looks like a very weird security hole (accessing Electron APIs unexpectedly) so I had it blocked by default. But thanks to this DI refactoring someone can bring back this behavior (opening links in an Electron window).

Reply:
Feel free to remove this code if you think it's for the best.